consensus/hotstuff/timeout_collector.go

```diff
@@ -0,0 +1,50 @@
+package hotstuff
+
+import (
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// OnTCCreated is a callback which will be used by TimeoutCollector to submit a TC when it's able to create it.
+// This callback is called from processing logic.
+type OnTCCreated func(tc *flow.TimeoutCertificate)
+
+// OnPartialTCCreated is a callback which will be used by TimeoutCollector to notify about collecting f+1 timeouts for
+// some view.
+// This callback is called from processing logic.
+type OnPartialTCCreated func(view uint64)
+
+// OnNewQCDiscovered is a callback which will be called to notify about new validated QC.
+// This callback is called from processing logic.
+type OnNewQCDiscovered func(*flow.QuorumCertificate)
+
+// OnNewTCDiscovered is a callback which will be called to notify about new validated TC.
+// This callback is called from processing logic.
+type OnNewTCDiscovered func(*flow.TimeoutCertificate)
+
+// TimeoutCollector collects all timeout objects for a specified view. On the happy path, it
+// generates a TimeoutCertificate when enough timeouts have been collected.
+type TimeoutCollector interface {
+	// AddTimeout adds a timeout object to the collector
+	// When f+1 TOs will be collected then callback for partial TC will be triggered,
+	// after collecting 2f+1 TOs a TC will be created and passed to the EventLoop.
```

Comment:
for all our implementations, we assume the max possible number of byzantine nodes. Using `f` to represent the number of malicious nodes is uncommon in our code base, in particular because nodes are weighted.

```suggestion
	// When TOs from strictly more than 1/3 of consensus participants (measured by weight)
	// were collected, the callback for partial TC will be triggered.
	// After collecting TOs from a supermajority, a TC will be created and passed to the EventLoop.
```